spec/models/shortened_url_spec.rb

```diff
@@ -123,31 +123,48 @@
 
       context "duplicate unique key" do
         let(:duplicate_key) { 'ABCDEF' }
-        it 'should try until it finds a non-dup key' do
-          Shortener::ShortenedUrl.where(unique_key: duplicate_key).delete_all
-          Shortener::ShortenedUrl.create!(url: Faker::Internet.url, custom_key: duplicate_key)
-          short_url = Shortener::ShortenedUrl.create!(url: Faker::Internet.url, custom_key: duplicate_key)
-          expect(short_url).not_to be_nil
-          expect(short_url.unique_key).not_to be_nil
-          expect(short_url.unique_key).not_to eq duplicate_key
-        end
-
-        context 'the persist_retries is set to 1' do
-          around do |example|
+        context 'without retry' do
+          around do |spec|
             tries = Shortener.persist_retries
-
             Shortener.persist_retries = 0
-            example.run
+            spec.run
             Shortener.persist_retries = tries
           end
 
-          it 'raises the non unique error' do
+          it 'finds a non-dup key' do
             Shortener::ShortenedUrl.where(unique_key: duplicate_key).delete_all
             Shortener::ShortenedUrl.create!(url: Faker::Internet.url, custom_key: duplicate_key)
-            expect {
-              Shortener::ShortenedUrl.create!(url: Faker::Internet.url, custom_key: duplicate_key)
-            }.to raise_error(ActiveRecord::RecordNotUnique)
+            short_url = Shortener::ShortenedUrl.create!(url: Faker::Internet.url, custom_key: duplicate_key)
+            expect(short_url).not_to be_nil
+            expect(short_url.unique_key).not_to be_nil
+            expect(short_url.unique_key).not_to eq duplicate_key
           end
+
+          it 'unscoped query to finds a non-dup key' do
+            Shortener::ShortenedUrl.where(unique_key: duplicate_key).delete_all
+            Shortener::ShortenedUrl.create!(url: Faker::Internet.url, custom_key: duplicate_key)
+            short_url = Shortener::ShortenedUrl.where(
+              url: Faker::Internet.url, category: :test
+            ).create!(url: Faker::Internet.url, custom_key: duplicate_key)
+            expect(short_url).not_to be_nil
+            expect(short_url.unique_key).not_to be_nil
+            expect(short_url.unique_key).not_to eq duplicate_key
+          end
+        end
+
+        it 'use retry in case with DB unique constraint exception' do
+          Shortener::ShortenedUrl.where(unique_key: duplicate_key).delete_all
+          Shortener::ShortenedUrl.create!(url: Faker::Internet.url, custom_key: duplicate_key)
+
+          query = double
+          query.stub(:exists?).and_return(false)
```

Comment:
shouldn't this be `true` to test the retry logic? Could this test be a false positive?

Reply:
`retry` is triggered when there is `ActiveRecord::RecordNotUnique` due concurrency, while query is used to make sure that token is uniq at list in current moment

what about `true`, I think it's related to order of stubs, so `receive(:unscoped).and_call_original`is called at first and then `receive(:unscoped).and_return(query).once` or maybe the stub of query is unnecessary at all